lib/api/objectPut.js

```diff
@@ -190,95 +179,71 @@ function objectPut(authInfo, request, streamingV4Params, log, callback) {
-                                            null, corsHeaders);
-                                    }
-                                    return _storeIt(bucket, objectKey,
-                                        objMD, authInfo, canonicalID,
-                                        cipherBundle, request,
-                                        streamingV4Params, log,
-                                            (err, contentMD5) =>
-                                                callback(err, contentMD5,
-                                                    corsHeaders));
-                                });
-                        }
-                        return _storeIt(bucket, objectKey, objMD,
-                                        authInfo, canonicalID, null, request,
-                                        streamingV4Params, log,
-                                (err, contentMD5) =>
-                                    callback(err, contentMD5, corsHeaders));
-                    });
-        }
-        if (serverSideEncryption) {
-            return kms.createCipherBundle(
-                serverSideEncryption,
-                log, (err, cipherBundle) => {
-                    if (err) {
-                        return callback(errors.InternalError, null,
-                            corsHeaders);
-                    }
-                    return _storeIt(bucket, objectKey, objMD,
-                                    authInfo, canonicalID, cipherBundle,
-                                    request, streamingV4Params, log,
-                            (err, contentMD5) =>
-                                callback(err, contentMD5, corsHeaders));
-                });
-        }
-        return _storeIt(bucket, objectKey, objMD, authInfo, canonicalID,
-                        null, request, streamingV4Params, log,
-                        (err, contentMD5) =>
-                            callback(err, contentMD5, corsHeaders));
+        return async.waterfall([
+            function handleTransientOrDeleteBuckets(next) {
+                if (bucket.hasTransientFlag() || bucket.hasDeletedFlag()) {
+                    return cleanUpBucket(bucket, canonicalID, log, next);
+                }
+                return next();
+            },
+            function createCipherBundle(next) {
+                const serverSideEncryption = bucket.getServerSideEncryption();
+                if (serverSideEncryption) {
+                    return kms.createCipherBundle(
+                            serverSideEncryption, log, next);
+                }
+                return next(null, null);
+            },
+            function objectCreateAndStore(cipherBundle, next) {
+                return createAndStoreObject(bucketName,
+                bucket, objectKey, objMD, authInfo, canonicalID, cipherBundle,
+                request, streamingV4Params, log, next);
+            },
+        ], (err, res) => {
```

Comment:
can we call this resInfo or something similar since it's not actually the request object?